### PR TITLE
Implement an in-memory Snapshot for MCP protocol

### PR DESCRIPTION
**Pull request by ozevren**

### CODE REVIEW

File: galley/pkg/mcp/snapshot/inmemory.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package snapshot
+
+import mcp "istio.io/api/config/mcp/v1alpha1"
+
+// InMemory Snapshot implementation
+type InMemory struct {
```
Comment (ayj) [inline galley/pkg/mcp/snapshot/inmemory.go]: This should be protected with a `sync.Mutex`.

File: galley/pkg/mcp/snapshot/inmemory.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	return s.resources[typ]
+}
+
+// Version is an implementation of Snapshot.Version
+func (s *InMemory) Version(typ string) string {
+	return s.versions[typ]
+}
+
+// Set the values for a given type.
+func (s *InMemory) Set(typ string, version string, resources []*mcp.Envelope) {
```
Comment (ayj) [inline galley/pkg/mcp/snapshot/inmemory.go]: Can you add a comment that snapshot is essentially immutable once passed to the cache and `Set()` should not longer be invoked?

File: galley/pkg/mcp/snapshot/inmemory.go (hunk 1 above)
Comment (ozevren) [reply galley/pkg/mcp/snapshot/inmemory.go]: Given that this is semantically immutable, do we need sync.Mutex?

Comment (ayj) [reply galley/pkg/mcp/snapshot/inmemory.go]: The current code suggests a mutable in-memory snapshot with no locking. We either ought to:

- add locking and allow it to be safely mutable, relying on convention for immutability after snapshot is set. An explicit `Freeze()` method could prevent mutation.

or 

- remove `Set()` and enforce immutability at object creation time, i.e. pass map of resources and verions to `New()`.

File: galley/pkg/mcp/snapshot/inmemory_test.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+package snapshot
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/gogo/protobuf/types"
+
+	"istio.io/api/config/mcp/v1alpha1"
```
Comment (ayj) [inline galley/pkg/mcp/snapshot/inmemory_test.go]: nit: can we use consistent package aliasing throughout, e.g. `mcp "istio.io/api/config/mcp/v1alpha1"`?

File: galley/pkg/mcp/snapshot/inmemory.go (hunk 1 above)
Comment (ozevren) [reply galley/pkg/mcp/snapshot/inmemory.go]: Added Freeze() and made Set() panic when frozen.

File: galley/pkg/mcp/snapshot/inmemory.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+// NewInMemory creates a new InMemory snapshot implementation
+func NewInMemory() *InMemory {
+	return &InMemory{
+		resources: make(map[string][]*mcp.Envelope),
+		versions:  make(map[string]string),
+	}
+}
+
+// Resources is an implementation of Snapshot.Resources
+func (s *InMemory) Resources(typ string) []*mcp.Envelope {
```
[the remark this answers is not in the document]
Comment (ozevren) [reply galley/pkg/mcp/snapshot/inmemory.go]: Either is fine. There is no room for confusion either way. If we're going to do it, let's do it in a separate check-in.

**Pull request merged**